rfcs/20210519-single-pass-bnorm/README.md

```diff
@@ -0,0 +1,126 @@
+where N = number of items in given channel in tensor.
+
+In the formula there's a subtraction of two large values which is vulnerable
+to catastrophic cancellation. Rounding errors in the two big numbers may be
+large in relation to result of the formula. In order to minimize rounding
+errors:
+- Sum and SumOfSquares will be stored in fp32 precision, regardless of input
+	data type
+- Kahan summation algorithm will be used
+
+The last line, "if var < 0 then var = 0" is there to make sure rounding errors
+won't cause variance to go negative. Further in batchnorm calculations square
+root is taken from variance. Negative value would result in NaN.
+
+New formula should be considered experimental. By default the old one will be
+used. There needs to be a way for user to choose formula. It is not decided yet
+how exactly user will toggle it.
+
+### Enable through environmental variable
+OneDNN will read env var `DNNL_BNORM_ONE_PASS`. If not set or equal to zero, 
+old formula will be used.
+
+Pros:
+- API/ABI compatibility
+- Simplest to implement and use
+- Clearly defined default state
+- If needed, may be kept confidential and not disclosed to external customes 
+until tested internally
+
+Cons:
+- No fine-tuning per instance of bnorm primitive
+- Harder to reproduce bugs related to bnorm: value of this flag will need to be
+ reported in bug reports to allow reproduction.
+ - Inconvenient long-term. State of oneDNN will have a dependency which is not
+ obvious to users.
+
+### Enable through additional flag in bnorm primitive descriptor
+Bnorm descriptor has flags field `dnnl_normalization_flags_t`. Add new flag 
+`dnnl_one_pass = 0x20u`. When set to non-zero, new formula will be used.
+
+Pros:
+- Fine-tuning per instance of bnorm primitive
+- Suitable to become permanent solution, should both formulas be kept in oneDNN
+
+Cons:
+- API change needed.
+- In case a customer was passing garbage into bit5 of flags, the default
+state is undefined and behavior may unexpectedly change.
+- If we decide to remove it later and keep only one formula, it will require
+another API change; it will leave behind a deprecated flag.
+- Inconsistency: not all implementations of bnorm primitive support that flag.
+Specifically, CPU implementation doesn't have it.
+- Inconvenient for validation. In order to test this feature with a topology,
+code change in frameworks in needed as frameworks will need to set this flag 
+for each bnorm instance.
+
+### Enable through new API call
+Introduce new API function `void set_bnorm_one_pass_mode(bool onepass)`. This
+function will set global variable which will subsequently control behavior of
+all instances of bnorm primitive.
+
+Pros:
+- API stays backwards-compatible
+- Clearly defined default state
+
+Cons:
+- API change needed.
+- No fine-tuning per instance of bnorm primitive
+- If we decide to remove it later and keep only  one formula, it will require
+ another API change; it will leave behind a deprecated function.
+
```

Comment:
Option 4. The feature you describe is a new algorithm for batch normalization, so I think alg_kind_t fits this feature perfectly:
```cpp
// dnnl_types.h
typedef enum {
...
dnnl_batch_normalization_direct,
dnnl_batch_normalization_one_pass, // name is still bad and doesn't tell how it is computed
...
} dnnl_alg_kind_t;

// dnnl.h
dnnl_status_t DNNL_API dnnl_batch_normalization_forward_desc_init_v2(
        dnnl_batch_normalization_desc_t *bnrm_desc, dnnl_prop_kind_t prop_kind, dnnl_alg_kind_t alg_kind,
        const dnnl_memory_desc_t *data_desc, float epsilon, unsigned flags);
```

Reply:
This would be nice if we were certain that both algorithms are going to stay. But we don't know that yet. Assuming it won't stay, then this becomes the most inconvenient solution for end users: one backwards-incompatible API change now and another one when we delete it.

Reply:
Overall, it seems that it would be similar to option 2, but worse :-)
- it would require to expose new op descriptor, new dispatch, ... and as @gtrzebia mentioned, there is no way to make one default, they would just cohabit next to each other
- FWK have apparently no way to know which one to chose with that granularity, so they would just pick one algorithm, and always use the same for all BN ops.

With the two above points in mind, I don't know if the new algorithm kind or flag is worth it.